minari/serialization.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+import json
+from typing import Dict, Union
+
+import gymnasium as gym
+import numpy as np
+
+
+def serialize_space(space: gym.spaces.Space, to_string=True) -> Union[Dict, str]:
+    if isinstance(space, gym.spaces.Box):
+        result = {}
+        result["type"] = "Box"
+        result["dtype"] = str(space.dtype)
+        result["shape"] = list(space.shape)
+        result[
+            "low"
+        ] = (
+            space.low.tolist()
+        )  # we have to use python float type to serialze the np.float32 types
+        result["high"] = space.high.tolist()
+    elif isinstance(space, gym.spaces.Discrete):
+        result = {}
+        result["type"] = "Discrete"
+        result["dtype"] = "int64"  # this seems to be hardcoded in Gymnasium
+        result["start"] = int(
+            space.start
+        )  # we need to cast from np.int64 to python's int type in order to serialize
+        result["n"] = int(space.n)
+    elif isinstance(space, gym.spaces.Dict):
+        result = {"type": "Dict", "subspaces": {}}
+        for key in space.spaces.keys():
+            result["subspaces"][key] = serialize_space(
+                space.spaces[key], to_string=False
+            )
+    elif isinstance(space, gym.spaces.Tuple):
+        result = {"type": "Tuple", "subspaces": []}
+        for subspace in space.spaces:
+            result["subspaces"].append(serialize_space(subspace, to_string=False))
+    if to_string:
+        return json.dumps(result)
+    else:
+        return result
+
```

Comment:
Missing an `else` branch
When you have an unsupported space, this will throw an uninformative error at line 43 (`result` is not initialized); better to have an informative error in the else branch

Reply:
Addressed, and also added a test to check that a TypeError is correctly raised if unsupported space types are serialized.